packages/slate-react/src/plugins/react/queries.js

```diff
@@ -431,10 +434,12 @@ function QueriesPlugin() {
     }
 
     const { document } = value
-    const range = document.createRange({
-      anchor,
-      focus,
-    })
+    const range = document
+      .createRange({
+        anchor,
+        focus,
+      })
+      .normalize(document, editor)
```

Comment:
@justinweiss hmm yeah good call. 

For the (a, "valid") case, that's the one I was thinking. I think it would be good if we can just ensure that Slate functions always return valid ranges, and then throw errors if people pass in invalid ones.

For the (b, "ambiguous") case, I hadn't thought about that as much. But it presents another issue... because even though we want to ensure there are no ambiguous cases, you might still need to construct the "unsanctioned" range for an interim state, or even when try to fix an ambiguous case? Such that it's less of a global "all ranges must be unambiguous". And more of a "the `Selection` must end up as unambiguous after each normalization pass of the editor".

For (a), it's not schema-specific, which is why we can remove it.

For (b), it is. But instead of having it at the `range.normalize`-type level (in the model). What if we added it at the `editor.normalizeSelection`-type level (in the controller) whenever operations are applied, to ensure that the selection stays in an unambiguous state. While still allow other ranges to refer to an ambiguous case if necessary.

(This might be already the approach you're going for! My heads a little muddled with just getting back into the different branches I've got going right now.)

Reply:
That makes sense to me. There are still a few cases where we want to force ambiguities to be removed, especially when handling events. For example, Safari _really_ wants cursors to be at the end of inlines, so even if the Slate data model says the selection is at the beginning of the next text node, `onBeforeInput` will insert the text into the inline when it calls `insertTextAtRange`. So those for sure need to be disambiguated. I think at the very least it makes sense to disambiguate any time we convert a browser selection to a Slate selection or finish a batch of operations. But I'm running into the same thing as I'm trying these changes out in our own app -- if I call a change method that puts a selection into the end of a node, _I want it to stay there_. At least until the next render pass.

I'll do another pass of this and see if it clears things up.

Reply:
OK, that seems to work nicely. I figured tying disambiguation to `normalizeDirtyNodes` felt right, and we can still clean up the selections we get from the browser in slate-react `queries.js`.

It also creates a few more deprecated functions we can just wholesale remove. `resolve*` can all go away. `resolveToTextNodes` / `resolveToTextNode` are just for raising if the selection looks weird, and making sure path / key are in sync -- we should be able to remove all those calls someday.

Reply:
Thinking about it more, maybe disambiguation should happen during `flush` instead? That way, someone could do something like `moveToEndOfNode().unwrapInline()` and not be confused about the cursor moving in between those two calls.

Reply:
@justinweiss I think they should be able to use `withoutNormalizing()` to do that.

Being able to remove all of the `resolve*` sounds amazing

Reply:
Playing around with it, delaying disambiguation until `flush` seems to work well. As a plugin author, you don't have to think about it. Selections start in the right place place, and in your event handlers, selections just go where you put them. You don't even really have to think about it.